lib/Dialect/FIRRTL/FIRRTLConnectionGraph.cpp

```diff
@@ -0,0 +1,222 @@
+//===- FIRRTLConnectionGraph.h - Graph of connections in FIRRTL -*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides utilities for working with FIRRTL operations using LLVM
+// graph utilties.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/FIRRTL/FIRRTLConnectionGraph.h"
+
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+
+using namespace circt;
+using namespace firrtl;
+using namespace detail;
+
+//===----------------------------------------------------------------------===//
+//
+// FModuleOpIterator
+//
+//===----------------------------------------------------------------------===//
+
+FModuleOpIterator::FModuleOpIterator(FModuleOp op) : op(op) {
+  auto &operations = op.getBodyBlock()->getOperations();
+  iterator = operations.begin();
+  iteratorEnd = operations.end();
+}
+
+bool FModuleOpIterator::operator==(const FModuleOpIterator &other) const {
+  if (iterator == iteratorEnd)
+    return other.iterator == other.iteratorEnd;
+
+  return iterator == other.iterator && iteratorEnd == other.iteratorEnd;
+}
+
+FModuleOpIterator &FModuleOpIterator::operator++() {
+  assert(iterator != iteratorEnd &&
+         "incrementing past the end of the iterator");
+  ++iterator;
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+//
+// FConnectLikeIterator
+//
+//===----------------------------------------------------------------------===//
+
+bool FConnectLikeIterator::operator==(const FConnectLikeIterator &other) const {
+  if (visited)
+    return other.visited;
+  return op == other.op && !other.visited;
+}
+
+FIRRTLOperation &FConnectLikeIterator::operator*() {
+  assert(!visited && "tried to dereference past the end");
+  auto dest = op.getDest();
+  FIRRTLOperation *result;
+  if (auto blockArg = dyn_cast<BlockArgument>(dest)) {
+    result = blockArg.getParentBlock()->getParentOp();
```

Comment:
How does this work ? Does it return the module op , so the iterator starts traversing all the module ops whenever a blockarg is the destination of a connect op ?